network/proto/senders.go

```diff
@@ -63,10 +72,32 @@ func (s defaultMessageSender) sendTransactionListQuery(peer p2p.PeerID, blockDat
 }
 
 func (s defaultMessageSender) sendTransactionList(peer p2p.PeerID, transactions []dag.Transaction, blockDate time.Time) {
-	envelope := createEnvelope()
+	if len(transactions) == 0 {
+		// messages are asynchronous so the requester is not waiting for a response
+		return
+	}
+	// When the DAG grows it the transactions might not fit in 1 network message (defined by p2p.MaxMessageSizeInBytes),
+	// so we have to split it up in pages. Protobuf can calculate the size of a message, so we can check whether it will fit in the max. message size.
+	// However, this can be quite expensive since it will marshal the message to the wire format (thus, marshalling will happen twice).
+	// Furthermore, we'd have to do this for every transaction we add to the message to find the max. number of TXs that
+	// will fit in the message.
+	// Since this sounds all horribly inefficient, we just calculate the size of a network message with 1 transaction,
+	// and use that to guesstimate the number of transactions that will fit in a message, decreased by a safety margin.
+	// Since transactions just contain metadata of the transaction (and not the payload itself), all transactions should
+	// serialize to more or less the same number of bytes, making the calculation safe.
+	transactionsPerMessage := s.getTransactionsPerMessage(transactions)
+	numberOfMessages := int(math.Ceil(float64(len(transactions)) / float64(transactionsPerMessage)))
 	tl := toNetworkTransactions(transactions)
-	envelope.Message = &transport.NetworkMessage_TransactionList{TransactionList: &transport.TransactionList{Transactions: tl, BlockDate: uint32(blockDate.Unix())}}
-	s.doSend(peer, &envelope)
+
+	for i := 0; i < numberOfMessages; i++ {
+		upper := (i + 1) * transactionsPerMessage
+		if upper > len(tl) {
+			upper = len(tl)
+		}
+		envelope := createEnvelope()
+		envelope.Message = &transport.NetworkMessage_TransactionList{TransactionList: &transport.TransactionList{Transactions: tl[i*transactionsPerMessage : upper], BlockDate: uint32(blockDate.Unix())}}
+		s.doSend(peer, &envelope)
+	}
```

Comment:
There are two issues with this:

1. We can calculate the size of a single message once by using an `init` function to set a global variable or something like that (instead of doing it each time)
2. Currently, we use the `NetworkMessage` protobuf type to calculate the size of a single transaction but this will be very inefficient as it will count the overhead of the `NetworkMessage` and `TransactionList` type multiple times instead of only once
3. We can't safely predict the size of the transaction data as the JWS payload has no maximum size. Even with an error margin this could go wrong. Maybe it would be better to either add a restriction on the JWS payload (so that we can calculate the maximum possible size of the transaction data).

Reply:
1. is now being done, but it requires a transaction to do. So it won't be possible at initialize.
2. is only done once, so?
3. our JWS payload only differs by the number of hashes in the `prev` header. The payload is always a SHA256

Reply:
I misread part of the code so `2` is definitely not an issue. But for `1` we could create a fake transaction and base the size on that right? It shouldn't really matter whether it's one of the actual transactions or a fake one as we already take an error margin into account.

And `3` still stands right? As we don't know the number of hashes in the `prev` header or do you think that it shouldn't differ that much?

Reply:
1. Currently it does but we could create a 'fake' transaction and base the size on that right? Or do you think that the size will differ too much from a real-world transaction?
2. I meant that it will over-estimate the size of a single transaction as `messageSizePerTX` (on line `127`) will include the overhead of a network message and transaction list structure. So if there are 10 transactions it will count this overhead ten times (which results in less transactions being sent per page) instead of only once. Or am I missing something?
3. Check!

Reply:
0. when we switch to didcomm, all will change anyway
1. could be, but will probably be more sensitive to bugs since we have to make some assumptions
2. so we're on the safe side twice! Currently the estimate is that we will send 500 TXs per page (instead of a 1000 or so). This should not be reached in the future when we select TXs based on lampart clocks. Also historical syncs for new nodes should only be done from trusted nodes in a streaming way (not through messages).

in all, it's temporary

Reply:
> We can calculate the size of a single message once by using an init function to set a global variable or something like that (instead of doing it each time)

We could do that, with some fake transaction. Would need some refactoring an maybe some clever way to assure the "guesstimate" fake-TX is as close to a real-world TX as possible. But I don't think this calculation (once per `TransactionList` sent back) is going to be the bottleneck. However, I could cache it?

> Currently, we use the NetworkMessage protobuf type to calculate the size of a single transaction but this will be very inefficient as it will count the overhead of the NetworkMessage and TransactionList type multiple times instead of only once

I figured the overhead is minimal since the protobuf definition (as text) is compiled to a TLV-like (tag-length-value), very efficient optimized format. Meaning that an additional (almost empty) type just adds a few bytes, which is a very small overhead when compared to the size of the text-based JWS. More overhead means also we're more safe.
And since an extra page because our guestimating algorithm isn't optimized won't be much slower: the same amount of data is sent, just in separate messages.  

> We can't safely predict the size of the transaction data as the JWS payload has no maximum size. Even with an error margin this could go wrong. Maybe it would be better to either add a restriction on the JWS payload (so that we can calculate the maximum possible size of the transaction data).

The JWS payload size is fixed, since it's the SHA256 hash of the actual data. The rest of the JWS (headers) is strictly spec'd, so there will be some variation in size but not too much. That's why there's an additonal safety margin (fill to 75% capacity).